### PR TITLE
Added custom message to ensure only one

### PR DESCRIPTION
**Pull request by James-OpenGamma**
## Summary by CodeRabbit

- **New Features**
  - Added a method to ensure no more than one matching element in a stream, with custom error messages for better clarity.

### CODE REVIEW

**Comment by jodastephen** [comment]
@coderabbit review

**Comment by jodastephen** [comment]
@coderabbitai review

**Comment by James-OpenGamma** [comment]
@jodastephen Pretty cool, a lot of noise though haha

I do like the summary it added to the PR description.

File: modules/collect/src/main/java/com/opengamma/strata/collect/Guavate.java [hunk 1]
```diff
@@ -682,6 +682,7 @@ public static <R> Predicate<R> not(Predicate<R> predicate) {
    *
    * @param <T>  the type of element in the stream
    * @return the operator
+   * @throws IllegalArgumentException if more than one element is present
```
[the remark this answers is not in the document]
Comment (jodastephen) [reply modules/collect/src/main/java/com/opengamma/strata/collect/Guavate.java:685]: This is not good feedback - there is already an `@throws` tag.

File: modules/collect/src/main/java/com/opengamma/strata/collect/Guavate.java [hunk 2]
```diff
@@ -690,6 +691,31 @@ public static <T> BinaryOperator<T> ensureOnlyOne() {
...
+   * @param message the message template for the {@link IllegalArgumentException} with "{}" placeholders
+   * @param args the arguments for the message
+   * @return the operator
+   * @throws IllegalArgumentException if more than one element is present
+   */
+  public static <T> BinaryOperator<T> ensureOnlyOne(String message, Object... args) {
+    return (a, b) -> {
+      throw new IllegalArgumentException(Messages.format(message, args));
+    };
+  }
```
[the remark this answers is not in the document]
Comment (jodastephen) [reply modules/collect/src/main/java/com/opengamma/strata/collect/Guavate.java:717]: Thanks. I don't consider varargs performance to be an issue here, as `Messages.format` will dominate the method's performance characteristics.

File: modules/collect/src/test/java/com/opengamma/strata/collect/GuavateTest.java [hunk 3]
```diff
@@ -424,6 +424,17 @@ public void test_ensureOnlyOne() {
...
+  @Test
+  public void test_ensureOnlyOne_withCustomMessage() {
+    String message = "Expected one letter but found multiple for date {}";
+    LocalDate arg = LocalDate.of(2024, 4, 24);
+    assertThat(Stream.empty().reduce(Guavate.ensureOnlyOne(message, arg))).isEqualTo(Optional.empty());
+    assertThat(Stream.of("a").reduce(Guavate.ensureOnlyOne(message, arg))).isEqualTo(Optional.of("a"));
+    assertThatIllegalArgumentException().isThrownBy(() -> Stream.of("a", "b")
+            .reduce(Guavate.ensureOnlyOne(message, arg)))
+        .withMessage(Messages.format(message, arg));
+  }
```
[the remark this answers is not in the document]
Comment (jodastephen) [reply modules/collect/src/test/java/com/opengamma/strata/collect/GuavateTest.java:436]: Thanks. In this repository, methods are globally defined as not accepting null unless explicitly specified. As such, there is generally no need for tests in this repository to check for null inputs.

**Comment by jodastephen** [comment]
@coderabbitai resolve

**Review by jodastephen — APPROVED** [review]
